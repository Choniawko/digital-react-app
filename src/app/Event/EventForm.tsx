import React, { useState } from "react"
import { connect } from "react-redux"
import { initEvent } from "../../fixtures/Event"
import { Action, RootState } from "../../store"
import { actionCreators, Event } from "../../store/Event"

type InputEvent = React.ChangeEvent<HTMLInputElement>
type ButtonEvent = React.FormEvent<HTMLFormElement>
export interface Props {
  event: Event
  postEvent: (payload: Event) => Action<Event>
}

const EventForm = ({ event, postEvent }: Props): JSX.Element => {
  const [eventForm, setEventForm] = useState(event.id ? event : initEvent)
  const handleInputChange = ({ target: { name, value } }: InputEvent) => {
    setEventForm({ ...eventForm, [name]: value })
  }
  const handleSubmit = (e: ButtonEvent) => {
    e.preventDefault()
    postEvent(eventForm)
  }
  return (
    <div>
      <form onSubmit={handleSubmit}>
        <div>
          <label>Name</label>
          <input
            type="text"
            name="name"
            value={eventForm.name}
            onChange={handleInputChange}
          />
        </div>
        <div>
          <label>Start date</label>
          <input
            type="text"
            name="startDate"
            value={eventForm.startDate.toLocaleString()}
            onChange={handleInputChange}
          />
        </div>
        <div>
          <label>End date</label>
          <input
            type="text"
            name="endDate"
            value={eventForm.endDate.toLocaleString()}
            onChange={handleInputChange}
          />
        </div>
        <div>
          <label>Description</label>
          <input
            type="text"
            name="description"
            value={eventForm.description}
            onChange={handleInputChange}
          />
        </div>
        <div>
          <label>Location</label>
          <input
            type="text"
            name="location"
            value={eventForm.location}
            onChange={handleInputChange}
          />
        </div>
        <button>Add new event</button>
      </form>
    </div>
  )
}

const mapStateToProps = ({ event: { event } }: RootState) => ({
  event,
})
const mapDispatchToProps = {
  postEvent: actionCreators.postEvent.create,
}
export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(EventForm)
