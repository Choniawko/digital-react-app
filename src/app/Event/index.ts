export { default as EventList } from "./EventList"
export * from "./EventList"
export { default as EventItem } from "./EventItem"
export { default as EventForm } from "./EventForm"
