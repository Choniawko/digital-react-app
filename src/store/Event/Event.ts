export type Event = {
    id: number
    name: string
    startDate: Date
    endDate: Date
    description: string
    location: string
    category: Category
}

export type Category = {
    id: number
    name: string
}

export type EventState = {
    events: ReadonlyArray<Event>
    event: Event | null
    count: number
    previos: number | null
    next: number | null
    error: string
}
