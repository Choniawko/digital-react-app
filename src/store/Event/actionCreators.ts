import { Event } from "."
import ActionCreator from "../actionCreator"

export default {
  getEvents: ActionCreator<null>("EVENT_GET_EVENTS"),
  getEvent: ActionCreator<number>("EVENT_GET_EVENT"),
  postEvent: ActionCreator<Event>("EVENT_POST_EVENT"),
  updateEvents: ActionCreator<Event[]>("EVENT_UPDATE_EVENTS"),
  updateEvent: ActionCreator<Event>("EVENT_UPDATE_EVENT"),
  setError: ActionCreator<string>("EVENT_SET_ERROR"),
}
