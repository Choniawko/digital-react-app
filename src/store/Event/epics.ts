import { combineEpics, ofType, Epic } from "redux-observable"
import { of } from "rxjs"
import { mergeMap, catchError } from "rxjs/operators"
import { apiUrl, ajax } from "../../common/services"
import { Action } from "../actionCreator"
import { actionCreators } from "."
import { RootState } from ".."

const { getEvents, getEvent, updateEvents, updateEvent, setError } = actionCreators

const getEventsEpic: Epic<Action<null>, Action<any>, RootState> = actions$ =>
  actions$.pipe(
    ofType(getEvents.type),
    mergeMap(() =>
      ajax.get(apiUrl("events")).pipe(
        mergeMap(({ response }) => of(updateEvents.create(response))),
        catchError(error => of(setError.create(error))),
      ),
    ),
  )
const getEventEpic: Epic<Action<number>, Action<any>, RootState> = actions$ =>
  actions$.pipe(
    ofType(getEvent.type),
    mergeMap(id =>
      ajax.get(apiUrl(`event/${id}`)).pipe(
        mergeMap(({ response }) => of(updateEvent.create(response))),
        catchError(error => of(setError.create(error))),
      ),
    ),
  )
export const epics = combineEpics(getEventsEpic, getEventEpic)
