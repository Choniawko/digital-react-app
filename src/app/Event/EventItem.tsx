import React from "react"
import { Event } from "../../store/Event"

export interface Props {
  event: Event
}
export default ({ event: { name, location, description } }: Props) => {
  return (
    <div>
      <div>{name}</div>
      <div>{location}</div>
      <div>{description}</div>
    </div>
  )
}
