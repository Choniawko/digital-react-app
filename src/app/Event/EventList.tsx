import React, { useEffect } from "react"
import { RouteComponentProps } from "react-router-dom"
import { Event, actionCreators } from "../../store/Event"
import { RootState } from "../../store"
import { connect } from "react-redux"

type ParamsProps = null
export interface Props extends RouteComponentProps<ParamsProps> {
    events: ReadonlyArray<Event>
    getEvents: () => void
}

const EventList = ({ events, getEvents }: Props): JSX.Element => {
    useEffect(() => {
        getEvents()
    }, [])
    return (
        <>
            <h3>EventList</h3>
            <div className="list">
                {events.map(({ id, name }) => (
                    <div key={id}>{name}</div>
                ))}
            </div>
        </>
    )
}

const mapStateToProps = ({ event: { events } }: RootState) => ({
    events,
})
const mapDispatchToProps = {
    getEvents: actionCreators.getEvents.create,
}
export default connect(
    mapStateToProps,
    mapDispatchToProps,
)(EventList)