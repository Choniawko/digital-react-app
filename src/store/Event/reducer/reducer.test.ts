import { eventReducer, actionCreators, INITIAL_STATE } from ".."
import { testEvent } from "../../../fixtures/Event"

describe("EventReducer", () => {
    it("calls updateEvents action", () => {
        const action = {
          type: actionCreators.updateEvents.type,
          payload: [testEvent],
        }
        expect(eventReducer(INITIAL_STATE, action)).toMatchSnapshot()
      })
      it("calls updateEvent action", () => {
        const action = {
          type: actionCreators.updateEvent.type,
          payload: testEvent,
        }
        expect(eventReducer(INITIAL_STATE, action)).toMatchSnapshot()
      })
});