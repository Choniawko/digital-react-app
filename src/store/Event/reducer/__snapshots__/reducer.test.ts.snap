// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventReducer calls updateEvent action 1`] = `
Object {
  "count": 0,
  "error": "",
  "event": Object {
    "category": Object {
      "id": 1,
      "name": "name",
    },
    "description": "description",
    "endDate": 2019-05-22T22:00:00.000Z,
    "id": 1,
    "location": "location",
    "name": "Name",
    "startDate": 2019-05-15T22:00:00.000Z,
  },
  "events": Array [],
  "next": null,
  "previos": null,
}
`;

exports[`EventReducer calls updateEvents action 1`] = `
Object {
  "count": 0,
  "error": "",
  "event": null,
  "events": Array [
    Object {
      "category": Object {
        "id": 1,
        "name": "name",
      },
      "description": "description",
      "endDate": 2019-05-22T22:00:00.000Z,
      "id": 1,
      "location": "location",
      "name": "Name",
      "startDate": 2019-05-15T22:00:00.000Z,
    },
  ],
  "next": null,
  "previos": null,
}
`;
